### PR TITLE
Modify init_from_list_of_dicts

### PR DESCRIPTION
**Pull request by dyq0811**
Fix for #5283. Will see how the tests go. @taldcroft

### CODE REVIEW

**Comment by pllim** [comment]
> Will see how the tests go

Tests failed. I recommend that you attempt to run the tests locally first and fix them before pushing out more commits. http://docs.astropy.org/en/stable/development/codeguide.html

**Comment by dyq0811** [comment]
@pllim Thanks! I will try the tests locally before commits.

**Comment by taldcroft** [comment]
@dyq0811 

First this needs to pass tests (locally), and then you need to write new tests.  One hint is that if you want to push a new commit for reviewers to see without starting the whole CI machinery, just put `[skip ci]` in the commit message at the end.

A minimal test that you can run locally (even just by hand in ipython during development) is something like:

```
vals = [{'a':1}, {'a': 2, 'b': 3}]
t = Table(vals)
assert t.colnames == ['a', 'b']
assert t.masked is True
assert np.all(t['a'] == [1, 2])
assert np.all(t['a'].mask == [False, False])
assert np.all(t['b'] == [-1, 3])  # The -1 value is ignored
assert np.all(t['b'].mask == [True, False])
assert t['a'].dtype.kind == 'i'
assert t['b'].dtype.kind == 'i'
```

**Comment by dyq0811** [comment]
Now I understand why I need first_vals for fixing data types of the columns... I was looking for placeholder with specified data type but didn't find any. So I replace them by some other values and mask them at the end. The current function passed the test for `vals = [{'a':1}, {'a': 2, 'b': 3}]` and `rows = [
 {'a': True, 'b': 42, 'c': 42.42, 'd': 'spam'}, # all fields present
 {           'b': 42, 'c': 42.42, 'd': 'spam'}, # missing bool `a`
 {'a': True,          'c': 42.42, 'd': 'spam'}, # missing int `b`
 {'a': True, 'b': 42,             'd': 'spam'}, # missing float `c`
 {'a': True, 'b': 42, 'c': 42.42             }, # missing string `d`
]` but still failed some tests.
`
def _init_from_list_of_dicts_2(self, data, names, dtype, n_cols, copy):
    """Initialize table from a list of dictionaries representing rows."""
    MISSING = object()
    names_from_data = set()
    for row in data:
        names_from_data.update(row)

    cols = {}
    missing_indexes = defaultdict(list)
    for name in names_from_data:
        cols[name] = []
        for i, row in enumerate(data):
            try:
                val = row[name]
            except KeyError:
                missing_indexes[name].append(i)
                val = MISSING
            cols[name].append(val)

    if missing_indexes:
        first_vals = {}
        for name in names_from_data:
            for val in cols[name]:
                if val is not MISSING:
                    first_vals[name] = val
                    break
    dtype = []
    for name in names_from_data:
        dtype.append(type(first_vals[name]))
        for index in missing_indexes[name]:
            cols[name][index] = first_vals[name]

    if all(name is None for name in names):
        names = sorted(names_from_data)
    n_cols = len(names)
    self._set_masked(True)
    self._init_from_dict(cols, names, dtype, n_cols, copy)

    if missing_indexes:
        for name, indexes in missing_indexes.items():
            self[name].mask[indexes] = True
    return
`

**Comment by dyq0811** [comment]
Got `ValueError: invalid literal for int() with base 10: 'int64'`.
Current code passed local tests for `vals = [{'a':1}, {'a': 2, 'b': 3}]` and `rows = [
 {'a': True, 'b': 42, 'c': 42.42, 'd': 'spam'},{           'b': 42, 'c': 42.42, 'd': 'spam'},{'a': True,          'c': 42.42, 'd': 'spam'}, {'a': True, 'b': 42,             'd': 'spam'}, {'a': True, 'b': 42, 'c': 42.42             }]`:
```
def _init_from_list_of_dicts(self, data, names, dtype, n_cols, copy):
    """Initialize table from a list of dictionaries representing rows."""
    MISSING = object()
    names_from_data = set()
    for row in data:
        names_from_data.update(row)

    cols = {}
    missing_indexes = defaultdict(list)
    for name in names_from_data:
        cols[name] = []
        for i, row in enumerate(data):
            try:
                val = row[name]
            except KeyError:
                missing_indexes[name].append(i)
                val = MISSING
            cols[name].append(val)

    first_vals = {}
    for name in names_from_data:
        for val in cols[name]:
            if val is not MISSING:
                first_vals[name] = val
                break
    dtype = []
    
    for name in names_from_data:
        dtype.append(type(first_vals[name]))
        if missing_indexes:
            for index in missing_indexes[name]:
                cols[name][index] = first_vals[name]

    if all(name is None for name in names):
        names = sorted(names_from_data)
    n_cols = len(names)
    self._set_masked(True)
    self._init_from_dict(cols, names, dtype, n_cols, copy)

    if missing_indexes:
        for name, indexes in missing_indexes.items():
            self[name].mask[indexes] = True
    return
```
Now I understand why I need first_vals... I was looking for placeholder with specified data type but didn't find any. So I replace them by first values and mask them at the end.

(Sorry the progress is delayed by school work)

File: astropy/table/table.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 import sys
 from collections import OrderedDict
+from collections import defaultdict
```
Comment (taldcroft) [inline astropy/table/table.py]: Put this in the same line with `OrderedDict` above.

**Comment by dyq0811** [comment]
@taldcroft (passed) local tests:
```
vals = [{'a':1}, {'a': 2, 'b': 3}]
t = Table(vals)
assert t.colnames == ['a', 'b']
assert t.masked is True
assert np.all(t['a'] == [1, 2])
assert np.all(t['a'].mask == [False, False])
assert np.all(t['b'] == [-10000, 3])
assert np.all(t['b'].mask == [True, False])
assert t['a'].dtype.kind == 'i'
assert t['b'].dtype.kind == 'i'
```
```
rows = [
 {'a': True, 'b': 42, 'c': 42.42, 'd': 'spam'}, # all fields present
 {           'b': 42, 'c': 42.42, 'd': 'spam'}, # missing bool `a`
 {'a': True,          'c': 42.42, 'd': 'spam'}, # missing int `b`
 {'a': True, 'b': 42,             'd': 'spam'}, # missing float `c`
 {'a': True, 'b': 42, 'c': 42.42             }, # missing string `d`
]

r = Table(rows)
assert r.colnames == ['a', 'b', 'c', 'd']
assert r.masked is True
assert np.all(r['b'] == [42, 42, -10000, 42, 42])
assert np.all(r['b'].mask == [False, False,  True, False, False])
assert np.all(r['d'].mask == [False, False, False, False, True])
assert r['a'].dtype.kind == 'b'
assert r['b'].dtype.kind == 'i'
assert r['c'].dtype.kind == 'f'
```
`Table(rows=[{'a':1}, {'a': 2, 'b': 3,'x':'s'},{'a': 2, 'b': 3.1},{'z':'k'},{'a':-10000},{'h':int}])`
Screen Shot 2019-04-07 at 11 36 23 AM

**Comment by dyq0811** [comment]
@taldcroft As the data type is not determined in the code of `init_from_list_of_dicts`, which part could be wrong?
Screen Shot 2019-04-14 at 2 42 26 PM

**Comment by taldcroft** [comment]
@dyq0811 - with apologies, I'm leaving for a 2-week vacation shortly.  I'll have a look when I return if you haven't figured it out by then!

**Comment by himanshupathak21061998** [comment]
@dyq0811  use 
```
assert t['a'].dtype.type == np.int64
```
Instead of 
```
assert t['a'].dtype.type == np.int32
```
Because you are using `dtype='i4'` which is `np.int64` not `np.int32`

**Comment by saimn** [comment]
> Because you are using dtype='i4' which is np.int64 not np.int32

No, `i4` is int32 (4 bytes).

@dyq0811 - I did not follow the PR so this is just a guess: the `dtype` passed to `_init_from_list_of_dicts` is not used in your code, so the dtype is guessed from the Python type, which leads to `int64` by default. You probably need to check if the `dtype` argument is not None, and use this at some point ?

**Comment by dyq0811** [comment]
@saimn Thanks, I think you are right. I end up using the condition `dtype == None or len(dtype) != n_cols` and the bug disappears.

The following error may just be fixed by commenting out the raise value error test. I am still looking for `table/construct_table.rst`....
Screen Shot 2019-04-15 at 7 06 49 PM
Local test:
Screen Shot 2019-04-15 at 7 07 00 PM

**Comment by taldcroft** [comment]
I want to evaluate if this is close enough to be fixed up for 4.0.

**Pull request closed without merging**